harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -389,11 +389,15 @@ def _train_for_step(self, step_id: int, batches_per_step: int) -> workload.Respo
             if self.use_amp():
                 with apex.amp.scale_loss(loss, self.context.optimizer) as scaled_loss:
                     scaled_loss.backward()
+                    if self.hvd_config.use and communicate_and_update:
+                        # When using horovod, we need to finish communicating gradient
+                        # updates before they are unscaled which happens when.
+                        self.context.optimizer.synchronize()
             else:
                 loss.backward()
 
             if communicate_and_update:
-                if self.hvd_config.use:
+                if self.hvd_config.use and not self.use_amp():
                     self.context.optimizer.synchronize()
```

Comment:
non-blocking: I slightly prefer having this line as part of the `else` above. Since it takes a bit of a careful reading to realize that we are not calling synchronize twice.

Reply:
Moved it around, still not a great layout but I do agree its bit easier to follow now.